index.js

```diff
@@ -141,6 +141,14 @@ function printMap(val, indent, prevIndent, refs, maxDepth, currentDepth, plugins
 }
 
 function printObject(val, indent, prevIndent, refs, maxDepth, currentDepth, plugins) {
+  if (val.toJSON) {
+    while (val.toJSON) {
+      val = val.toJSON();
+    }
+    if (typeof val === 'string') {
+      return /(\\|\"|\')/.test(val) ? print(val) : val;
```

Comment:
I don't want to wrap the whole thing in quotes if I don't have to. Consider:

```js
print({foo: {toJSON: () => "[Internal Object]"}});
// should print:
Object {
  foo: [Internal Object]
}
// instead of
Object {
  foo: "[Internal Object]"
}
// but when escape characters are part of it, we should print it as a real string:
print({foo: {toJSON: () => "['Internal' Object]"}});
Object {
  foo: "[\\\'Internal\\\' Object]"
}
```

Does this make sense? This is similar to how we quote object literal keys in JS if they wouldn't be valid JS otherwise.